libs/s25main/ingameWindows/iwEconomicProgress.cpp

```diff
@@ -0,0 +1,200 @@
+#include "gameData/const_gui_ids.h"
+
+iwEconomicProgress::iwEconomicProgress(const GameWorldViewer& gwv)
+    : IngameWindow(CGI_ECONOMICPROGRESS, IngameWindow::posLastOrCenter,
+                   Extent(240, 96 + 26 * EconomyModeHandler::numGoodTypesToCollect), _("Economic Progress"),
+                   LOADER.GetImageN("resource", 41)),
+      gwv(gwv)
+{
+    const unsigned int textcolor[] = {COLOR_GREEN, COLOR_YELLOW, COLOR_RED};
+
+    const GameWorldBase& world = gwv.GetWorld();
+
+    EconomyModeHandler* eH = world.econHandler;
+
+    headline =
+      AddText(1, DrawPoint(11 + 27, 46), _("Player"), COLOR_GREEN, FontStyle::LEFT | FontStyle::BOTTOM, NormalFont);
+
+    const std::vector<EconomyModeHandler::econTeam>& teams = eH->GetTeams();
+    unsigned int num_teams = teams.size();
+
+    // Teamreihenfolge bestimmen (eigenes Team zuerst)
+    unsigned int mainTeam = 0;
+    for(unsigned int i = 0; i < teams.size(); i++)
+    {
+        if(teams[i].inTeam(gwv.GetPlayer().GetPlayerId()))
+        {
+            mainTeam = i;
+            break;
+        }
+    }
+    teamorder.push_back(mainTeam);
+    for(unsigned int i = 0; i < teams.size(); i++)
+    {
+        if(i != mainTeam)
+        {
+            teamorder.push_back(i);
+        }
+    }
+    // Eventuell Fenster vergroessern
+    if(num_teams > 2)
+    {
+        Extent size = this->GetSize();
+        size.x += (num_teams - 2) * 63;
+        this->Resize(size);
+    }
+    // Die Warentabelle
+    const Extent btSize(26, 26);
+    for(unsigned int i = 0; i < eH->numGoodTypesToCollect; i++)
+    {
+        GoodType good = eH->GetTypes()[i];
+
+        const DrawPoint btPos(11, 48 + 26 * i);
+        AddImage(100 + i, btPos + btSize / 2, LOADER.GetMapImageN(2298), _(WARE_NAMES[good]));
+        const DrawPoint warePos = btPos + btSize / 2;
+        AddImage(200 + i, warePos, LOADER.GetMapImageN(WARES_TEX_MAP_OFFSET + good));
+
+        for(unsigned j = 0; j < 1 + num_teams; j++)
+        {
+            const DrawPoint txtPos = btPos + DrawPoint(26 + 63 * j, 0);
+            AddTextDeepening(300 + 10 * j + i, txtPos, Extent(63, btSize.y), TC_GREY, "?", NormalFont,
+                             textcolor[j < 2 ? j : 2]);
+        }
+    }
+
+    // Spielzeit-Fortschrittsanzeige
+    AddText(2, DrawPoint(11 + 30, 240), _("Percent elapsed:"), COLOR_ORANGE, FontStyle::LEFT | FontStyle::TOP,
+            NormalFont);
+    elapsedTime =
+      AddText(3, DrawPoint(30 + 63 * 3, 240), "%", COLOR_ORANGE, FontStyle::RIGHT | FontStyle::TOP, NormalFont);
+
+    // Hilfe-Button
+    AddImageButton(25, DrawPoint(11, 235), Extent(26, 26), TC_GREY, LOADER.GetImageN("io", 225), _("Help"));
+}
+
+iwEconomicProgress::~iwEconomicProgress() = default;
+
+void iwEconomicProgress::Draw_()
+{
+    IngameWindow::Draw_();
+
+    // Teamfarben zeichnen
+    const std::vector<EconomyModeHandler::econTeam>& teams = gwv.GetWorld().econHandler->GetTeams();
+    for(unsigned int t = 0; t < teams.size(); t++)
```

Comment:
Maybe: `for(const unsigned teamIdx: teamOrder)`?

Reply:
Here I also use the counter t to do the positioning of the rectangles so changing this would need to store the positioning delta together with the team order.

Reply:
You mean in `DrawPoint drawPt = GetDrawPos() + DrawPoint(37 + (t + 1) * 63, 22);`?
Then you could do `DrawPoint curTeamRectPos= GetDrawPos() + DrawPoint(37 + 63, 22);` before the loop and `curTeamRectPos.x += 63;` inside the loop, couldn't you?

Reply:
I didn't think of this. For a moment I thought this was some insanely clever idea I missed until now. Then I realized that curTeamRectPos.x basically becomes a counter variable. Still looks a bit cleaner, I agree. Thanks.

Also this made me notice that I use way too many explicit numbers in the construction of this window, so I cleaned that up.

Also made me realize that I can simply store pointers in teamOrder which makes it even simpler